File: neetpg-desktop/src/main/syncer.ts
import { join } from 'path'
import type { AppConfig, SyncStatus, SessionLog, SRState, TopicScores } from '../shared/types'
import { RepoSync } from './repo-sync'
import { CardCache } from './cache'
import { SREngine } from './sr-engine'
import { LLMService } from './llm-service'
import { parseMistakeFile, deriveHeading, deriveBullets } from './file-parser'
import { buildMistakeMarkdown } from './md-builder'

const SR_STATE_PATH = 'progress/sr-state.json'

export class Syncer {
  private status: SyncStatus = {
    lastSync: null,
    lastError: null,
    inProgress: false,
    totalCards: 0,
    phase: 'idle',
    parseErrors: []
  }
  private sink: ((s: SyncStatus) => void) | null = null

  constructor(
    private cfg: () => AppConfig,
    private repo: RepoSync,
    private cache: CardCache,
    private sr: SREngine,
    private llm: LLMService
  ) {}

  /** Attach a callback that receives every status transition — used by index.ts
   *  to push live updates to the renderer via IPC. */
  setStatusSink(fn: (s: SyncStatus) => void): void {
    this.sink = fn
  }

  private emit(patch: Partial<SyncStatus>): void {
    Object.assign(this.status, patch)
    this.sink?.({ ...this.status, totalCards: this.cache.allCards().length })
  }

  getStatus(): SyncStatus {
    return { ...this.status, totalCards: this.cache.allCards().length }
  }

  async sync(): Promise<{ changed: number; error: string | null }> {
    if (this.status.inProgress) return { changed: 0, error: null }
    this.emit({ inProgress: true, lastError: null, phase: 'listing' })
    let changed = 0
    const parseErrors: { path: string; reason: string }[] = []

    try {
      const tree = await this.repo.listTree()

      const mistakeFiles = tree.filter(
        (e) => e.path.startsWith('mistakes/') && e.path.endsWith('.md')
      )
      const sessionFiles = tree.filter(
        (e) => e.path.startsWith('sessions/') && e.path.endsWith('.json')
      )

      this.emit({ phase: 'fetching' })

      // Pull-merge remote SR state BEFORE anything else. Another device (or a
      // previous install) may have graded cards; per-card newest-wins merge
      // means progress survives reinstalls and multi-device use.
      const srEntry = tree.find((e) => e.path === SR_STATE_PATH)
      if (srEntry && this.cache.getBlobSha(SR_STATE_PATH) !== srEntry.sha) {
        try {
          const raw = await this.repo.fetchFile(SR_STATE_PATH)
          const remote = JSON.parse(raw) as SRState
          if (remote && remote.cards) this.sr.mergeRemote(remote)
          this.cache.setBlobSha(SR_STATE_PATH, srEntry.sha)
        } catch {
          // Bad/unreachable remote SR state — local store stays authoritative.
        }
      }

      for (const f of mistakeFiles) {
        if (this.cache.getBlobSha(f.path) === f.sha) continue
        try {
          const raw = await this.repo.fetchFile(f.path)
          const { card, error } = parseMistakeFile(raw, f.path, f.sha)
          if (card) {
            this.cache.upsertCard(card)
            changed++
          } else if (error) {
            // Record the sha so we don't refetch a permanently-broken file
            // every cycle, but surface it in the UI instead of dropping it.
            this.cache.setBlobSha(f.path, f.sha)
            parseErrors.push({ path: f.path, reason: error })
          }
        } catch {
          // Network hiccup on this file; try again next cycle.
        }
      }

      for (const f of sessionFiles) {
        if (this.cache.getBlobSha(f.path) === f.sha) continue
        try {
          const raw = await this.repo.fetchFile(f.path)
          const s = JSON.parse(raw) as Partial<SessionLog>
          if (s.sessionId) {
            this.cache.upsertSession(normalizeSession(s))
            this.cache.setBlobSha(f.path, f.sha)
            changed++
          }
        } catch {
          // ignore malformed session file
        }
      }

      this.joinSRState()
      this.cache.save()

      const c = this.cfg()
      if (this.llm.configured && c.enableMnemonics) {
        this.emit({ phase: 'enriching' })
        await this.enrichMissing()
      }

      if (c.githubPat) {
        this.emit({ phase: 'pushing' })
        await this.pushProgress()
      }

      this.emit({
        inProgress: false,
        lastSync: new Date().toISOString(),
        phase: 'done',
        parseErrors
      })
      return { changed, error: null }
    } catch (e) {
      // Sync failed (offline, rate-limited, …) — the local cache stays intact
      // and the renderer keeps serving cards from it; only the status changes.
      const msg = (e as Error).message
      this.emit({ inProgress: false, lastError: msg, phase: 'error', parseErrors })
      return { changed, error: msg }
    }
  }

  private joinSRState(): void {
    const sr = this.cache.allSR()
    for (const card of this.cache.allCards()) {
      const s = sr[card.id]
      if (s) {
        card.srStatus = s.status
        card.nextReview = s.nextReview
      }
      this.cache.upsertCard(card)
    }
  }

  private async enrichMissing(): Promise<void> {
    const missing = this.cache
      .allCards()
      .filter((c) => c.question && c.options.length > 0 && (!c.keyFact || !c.whyWrong))
      .slice(0, 5)
    const canPush = !!this.cfg().githubPat
    for (const card of missing) {
      const enriched = await this.llm.enrichCard(card)
      if (!enriched) continue
      card.keyFact = enriched.keyFact || card.keyFact
      card.whyWrong = enriched.whyWrong || card.whyWrong
      if (enriched.errorType) card.errorType = enriched.errorType as never
      if (enriched.tags.length) card.tags = enriched.tags
      card.factHeading = deriveHeading(card.topic, card.keyFact)
      card.factPoints = deriveBullets(card.keyFact)

      // Write the enrichment back to the repo so it reaches every device and
      // survives cache clears (the Android placeholder promises exactly this).
      if (canPush && card.filePath) {
        try {
          const sha = await this.repo.putFile(
            card.filePath,
            buildMistakeMarkdown(card),
            `enrich: ${card.id} key fact + analysis`
          )
          if (sha) card.lastModified = sha
        } catch {
          // Enrichment stays local this cycle; write-back retries next time
          // the card is still missing remote enrichment.
        }
      }
      this.cache.upsertCard(card)
    }
    this.cache.save()
  }

  private async pushProgress(): Promise<void> {
    try {
      const srState = this.sr.exportState()
      const srSha = await this.repo.putFile(
        SR_STATE_PATH,
        JSON.stringify(srState, null, 2),
        'sr: desktop review update'
      )
      // Record what we just wrote so next cycle's pull-merge skips it.
      if (srSha) this.cache.setBlobSha(SR_STATE_PATH, srSha)
      const topicScores = this.computeTopicScores()
      await this.repo.putFile(
        'progress/topic-scores.json',
        JSON.stringify(topicScores, null, 2),
        'progress: desktop topic scores'
      )
    } catch (e) {
      this.status.lastError = `push: ${(e as Error).message}`
    }
  }

  private computeTopicScores(): TopicScores {
    const map: Record<string, { unresolved: number; total: number }> = {}
    for (const c of this.cache.allCards()) {
      const key = c.subject
      if (!map[key]) map[key] = { unresolved: 0, total: 0 }
      map[key].total += 1
      if (!c.isResolved) map[key].unresolved += 1
    }
    const scores: Record<string, number> = {}
    for (const [k, v] of Object.entries(map)) {
      scores[k] = v.total > 0 ? Number((v.unresolved / v.total).toFixed(3)) : 0
    }
    return { lastUpdated: new Date().toISOString(), scores }
  }
}

function normalizeSession(s: Partial<SessionLog>): SessionLog {
  return {
    sessionId: s.sessionId || '',
    type: s.type || 'practice',
    startedAt: s.startedAt || '',
    endedAt: s.endedAt || '',
    totalQuestions: s.totalQuestions || 0,
    correct: s.correct || 0,
    wrong: s.wrong || 0,
    skipped: s.skipped || 0,
    scorePercent: s.scorePercent || 0,
    subjectBreakdown: s.subjectBreakdown || []
  }
}

export const progressPath = (root: string): string => join(root, 'progress')

File: neetpg-desktop-tauri/src-tauri/src/syncer.rs
//! One full sync cycle: list tree → pull-merge SR state → fetch & parse new
//! mistake/session files → LLM-enrich + write back → push progress. Every
//! status transition is emitted to the frontend as a `sync-status` event.

use crate::config::Config;
use crate::llm;
use crate::md_builder::build_mistake_markdown;
use crate::models::{now_iso, ParseError, SRState, SessionLog, SyncStatus, TopicScores};
use crate::parser::{derive_bullets, derive_heading, parse_mistake_file};
use crate::repo::Repo;
use crate::sr;
use crate::state::App;
use tauri::{AppHandle, Emitter};

const SR_STATE_PATH: &str = "progress/sr-state.json";

pub struct SyncResult {
    pub changed: usize,
    pub error: Option<String>,
}

async fn emit_status(app: &App, handle: &AppHandle, patch: impl FnOnce(&mut SyncStatus)) {
    let status = {
        let mut st = app.status.lock().unwrap();
        patch(&mut st);
        st.clone()
    };
    let mut status = status;
    status.total_cards = app.cache.lock().await.data.cards.len();
    let _ = handle.emit("sync-status", &status);
}

pub async fn sync(app: &App, handle: &AppHandle) -> SyncResult {
    if app
        .sync_running
        .swap(true, std::sync::atomic::Ordering::SeqCst)
    {
        return SyncResult { changed: 0, error: None };
    }
    let result = run(app, handle).await;
    app.sync_running.store(false, std::sync::atomic::Ordering::SeqCst);
    result
}

async fn run(app: &App, handle: &AppHandle) -> SyncResult {
    let cfg = app.config.lock().unwrap().clone();
    emit_status(app, handle, |st| {
        st.in_progress = true;
        st.last_error = None;
        st.phase = "listing".into();
    })
    .await;

    let mut changed = 0usize;
    let mut parse_errors: Vec<ParseError> = Vec::new();

    let outcome: Result<(), String> = async {
        let repo = Repo { http: &app.http, cfg: &cfg };
        let tree = repo.list_tree().await?;

        emit_status(app, handle, |st| st.phase = "fetching".into()).await;

        // Pull-merge remote SR state before anything else.
        if let Some(entry) = tree.iter().find(|e| e.path == SR_STATE_PATH) {
            let known = app.cache.lock().await.data.blob_shas.get(SR_STATE_PATH).cloned();
            if known.as_deref() != Some(&entry.sha) {
                if let Ok(raw) = repo.fetch_file(SR_STATE_PATH).await {
                    if let Ok(remote) = serde_json::from_str::<SRState>(&raw) {
                        let mut cache = app.cache.lock().await;
                        sr::merge_remote(&mut cache, &remote);
                        cache.data.blob_shas.insert(SR_STATE_PATH.into(), entry.sha.clone());
                    }
                }
            }
        }

        // Mistake files.
        for f in tree.iter().filter(|e| e.path.starts_with("mistakes/") && e.path.ends_with(".md")) {
            let known = app.cache.lock().await.data.blob_shas.get(&f.path).cloned();
            if known.as_deref() == Some(&f.sha) {
                continue;
            }
            let Ok(raw) = repo.fetch_file(&f.path).await else { continue };
            match parse_mistake_file(&raw, &f.path, &f.sha) {
                Ok(card) => {
                    app.cache.lock().await.upsert_card(card);
                    changed += 1;
                }
                Err(reason) => {
                    // Record the sha so a permanently-broken file isn't
                    // refetched every cycle — but surface it, never drop it.
                    let mut cache = app.cache.lock().await;
                    cache.data.blob_shas.insert(f.path.clone(), f.sha.clone());
                    parse_errors.push(ParseError { path: f.path.clone(), reason });
                }
            }
        }

        // Session logs.
        for f in tree.iter().filter(|e| e.path.starts_with("sessions/") && e.path.ends_with(".json")) {
            let known = app.cache.lock().await.data.blob_shas.get(&f.path).cloned();
            if known.as_deref() == Some(&f.sha) {
                continue;
            }
            let Ok(raw) = repo.fetch_file(&f.path).await else { continue };
            if let Ok(s) = serde_json::from_str::<SessionLog>(&raw) {
                if !s.session_id.is_empty() {
                    let mut cache = app.cache.lock().await;
                    cache.data.sessions.insert(s.session_id.clone(), s);
                    cache.data.blob_shas.insert(f.path.clone(), f.sha.clone());
                    changed += 1;
                }
            }
        }

        // Join SR fields onto cards, then persist.
        {
            let mut cache = app.cache.lock().await;
            let ids: Vec<String> = cache.data.cards.keys().cloned().collect();
            for id in ids {
                if let Some(s) = cache.data.sr.get(&id).cloned() {
                    if let Some(card) = cache.data.cards.get_mut(&id) {
                        card.sr_status = s.status;
                        card.next_review = s.next_review;
                    }
                }
            }
            cache.save();
        }

        // LLM enrichment + write-back.
        if !cfg.groq_api_key.is_empty() && cfg.enable_mnemonics {
            emit_status(app, handle, |st| st.phase = "enriching".into()).await;
            enrich_missing(app, &cfg).await;
        }

        // Push progress.
        if !cfg.github_pat.is_empty() {
            emit_status(app, handle, |st| st.phase = "pushing".into()).await;
            push_progress(app, &cfg).await;
        }

        Ok(())
    }
    .await;

    match outcome {
        Ok(()) => {
            emit_status(app, handle, |st| {
                st.in_progress = false;
                st.last_sync = Some(now_iso());
                st.phase = "done".into();
                st.parse_errors = parse_errors.clone();
            })
            .await;
            SyncResult { changed, error: None }
        }
        Err(msg) => {
            // Sync failed (offline, rate-limited, …) — the local cache stays
            // intact and the UI keeps serving cards from it.
            emit_status(app, handle, |st| {
                st.in_progress = false;
                st.last_error = Some(msg.clone());
                st.phase = "error".into();
                st.parse_errors = parse_errors.clone();
            })
            .await;
            SyncResult { changed, error: Some(msg) }
        }
    }
}

async fn enrich_missing(app: &App, cfg: &Config) {
    let missing: Vec<_> = {
        let cache = app.cache.lock().await;
        cache
            .data
            .cards
            .values()
            .filter(|c| {
                if c.question.is_empty() || c.options.is_empty() {
                    return false;
                }
                // Missing enrichment — always queue.
                if c.key_fact.is_empty() || c.why_wrong.is_empty() {
                    return true;
                }
                // Old verbose bullets (>50 chars each) — re-enrich with the
                // short-phrase prompt so the ambient slide gets crisp bullets.
                c.fact_points.is_empty() || c.fact_points.iter().any(|p| p.len() > 50)
            })
            .take(5)
            .cloned()
            .collect()
    };
    let can_push = !cfg.github_pat.is_empty();

    for mut card in missing {
        let Some(enriched) = llm::enrich_card(&app.http, &cfg.groq_api_key, &card).await else {
            continue;
        };
        if !enriched.key_fact.is_empty() {
            card.key_fact = enriched.key_fact;
        }
        if !enriched.why_wrong.is_empty() {
            card.why_wrong = enriched.why_wrong;
        }
        if matches!(enriched.error_type.as_str(), "conceptual" | "recall" | "silly") {
            card.error_type = enriched.error_type;
        }
        if !enriched.tags.is_empty() {
            card.tags = enriched.tags;
        }
        card.fact_heading = derive_heading(&card.topic, &card.key_fact);
        card.fact_points = if !enriched.fact_points.is_empty() {
            enriched.fact_points
        } else {
            derive_bullets(&card.key_fact, &card.fact_heading)
        };

        // Write the enrichment back to the repo so it reaches every device
        // and survives cache clears.
        if can_push && !card.file_path.is_empty() {
            let repo = Repo { http: &app.http, cfg };
            if let Ok(Some(sha)) = repo
                .put_file(
                    &card.file_path,
                    build_mistake_markdown(&card).as_bytes(),
                    &format!("enrich: {} key fact + analysis", card.id),
                )
                .await
            {
                card.last_modified = sha;
            }
        }
        app.cache.lock().await.upsert_card(card);
    }
    app.cache.lock().await.save();
}

async fn push_progress(app: &App, cfg: &Config) {
    let repo = Repo { http: &app.http, cfg };
    let (sr_state, topic_scores) = {
        let cache = app.cache.lock().await;
        (sr::export_state(&cache), compute_topic_scores(&cache))
    };

    if let Ok(body) = serde_json::to_vec_pretty(&sr_state) {
        if let Ok(Some(sha)) = repo.put_file(SR_STATE_PATH, &body, "sr: desktop review update").await {
            // Record what we just wrote so next cycle's pull-merge skips it.
            app.cache.lock().await.data.blob_shas.insert(SR_STATE_PATH.into(), sha);
        }
    }
    if let Ok(body) = serde_json::to_vec_pretty(&topic_scores) {
        let _ = repo
            .put_file("progress/topic-scores.json", &body, "progress: desktop topic scores")
            .await;
    }
}

fn compute_topic_scores(cache: &crate::cache::Cache) -> TopicScores {
    let mut map: std::collections::HashMap<String, (i64, i64)> = std::collections::HashMap::new();
    for c in cache.data.cards.values() {
        let e = map.entry(c.subject.clone()).or_insert((0, 0));
        e.1 += 1;
        if !c.is_resolved {
            e.0 += 1;
        }
    }
    TopicScores {
        last_updated: now_iso(),
        scores: map
            .into_iter()
            .map(|(k, (u, t))| (k, if t > 0 { (u as f64 / t as f64 * 1000.0).round() / 1000.0 } else { 0.0 }))
            .collect(),
    }
}
